### PR TITLE
BUG: groupby.transform(name) validates name is an aggregation

### LINKED ISSUE #14274: bug when filling missing values with transform?
Hello there,

Consider this

```
df = pd.DataFrame({'group' : ['A', 'A', 'A', 'B',
                                              'B', 'B', 'B', 'B'],
                                              'B' : [np.nan,np.nan,np.nan,-4,-2,5,8,7],
                                              'C' : [-5,5,-20,0,np.nan,5,4,-4]})

df
Out[13]: 
     B     C group
0  NaN  -5.0     A
1  NaN   5.0     A
2  NaN -20.0     A
3 -4.0   0.0     B
4 -2.0   NaN     B
5  5.0   5.0     B
6  8.0   4.0     B
7  7.0  -4.0     B
```

Now I want to **fill forward** the missing values in `C` for each group 

```

df.groupby('group').C.fillna(method ='ffill')
Out[11]: 
0    -5.0
1     5.0
2   -20.0
3     0.0
4     0.0
5     5.0
6     4.0
7    -4.0
Name: C, dtype: float64

df.groupby('group').C.transform('ffill')
Out[12]: 
0   -5.0
1   -5.0
2   -5.0
3    5.0
4    5.0
5    5.0
6    5.0
7    5.0
dtype: float64
```

the transform output is wrong. 
Is that expected? Pandas 18.1

### LINKED ISSUE #19354: Shortcut functions in transform are not grouped
```ipython
$ ipython3
Python 3.6.3 (default, Oct  3 2017, 21:45:48) 
Type 'copyright', 'credits' or 'license' for more information
IPython 6.2.1 -- An enhanced Interactive Python. Type '?' for help.

In [1]: import pandas as pd

In [2]: df = pd.DataFrame({'A': [1, 1, 2], 'B': [1, 2, 3]})

In [3]: df.groupby('A').transform('rank')
Out[3]: 
   B
0  1
1  1
2  2

In [4]: df.groupby('A').transform(lambda x: x.rank())
Out[4]: 
     B
0  1.0
1  2.0
2  1.0

```
#### Problem description
It seems like the functions provided through shorthand strings, such as 'rank', do not obey the groupings of data. For example, one would expect Out[3] and Out[4] to be the same in the code above. Instead, it seems like using `.transform('rank')` doesn't actually obey the grouping and just ranks independently. I've reproduced this with larger dataframes as well.

### LINKED ISSUE #22509: Wrong output of GroupBy transform with string input (e.g., transform('rank'))
#### Code Sample, a copy-pastable example if possible

```python
import pandas as pd

df = pd.DataFrame([[1, 10], [1, 20], [2, 40], [2, 30]], columns=['a', 'b'])

print("%s\n" % df)

print("Output of df.groupby('a')['b'].transform('rank'):\n%s\n" %
        df.groupby('a')['b'].transform('rank'))

print("Output of df.groupby('a')['b'].rank():\n%s\n" %
        df.groupby('a')['b'].rank())

print("Output of df.groupby('a')['b'].transform(lambda x: x.rank()):\n%s\n" %
        df.groupby('a')['b'].transform(lambda x: x.rank()))

print("Output of df.groupby('a')['b'].transform('cumcount'):\n%s\n" %
        df.groupby('a')['b'].transform('cumcount'))

print("Output of df.groupby('a')['b'].cumcount():\n%s\n" %
        df.groupby('a')['b'].cumcount())

```
#### Problem description

For simplicity, I will explain the issue for `SeriesGroupBy`, though the bug is present in `DataFrameGroupBy` as well. 

When calling `transform` on a `SeriesGroupBy` object with a string input `'string_input'` (e.g., `.transform('mean')`), the relevant code inside [`pandas/pandas/core/groupby/generic.py`](https://github.com/pandas-dev/pandas/blob/master/pandas/core/groupby/generic.py) will end up calling `SeriesGroupBy._transform_fast` on the function 
```
func = getattr(SeriesGroupBy, 'string_input')
```
(unless `'string_input'` is inside `base.cython_transforms`, currently consisting of `['cumprod', 'cumsum', 'shift', 'cummin', 'cummax']`). Inside `_transform_fast`, the result of applying `func` to the `SeriesGroupBy` object is then broadcast to the entire index of the original object. This works as expected if `func` returns a single value per group in the `GroupBy` object (e.g., for functions like `'mean'`, `'std'`, etc.). However, for functions like `rank`, `cumcount`, etc., that return several values per group, the result of broadcasting is nonsensical to my best knowledge.

Note: Index broadcasting works correctly if `transform` is called with a (non-cython) function, for example `lambda x: x.rank()`. In that case, `_fast_transform` is never called and the result is a simple concatenation of the results for each group.

#### Expected Output

The results of `df.groupby('a')['b'].transform('rank')` and `df.groupby('a')['b'].rank()` should be identical. Same for 'cumcount' and maybe other `GroupBy` functions.

#### Output of ``pd.show_versions()``

<details>

[paste the output of ``pd.show_versions()`` here below this line]
INSTALLED VERSIONS
------------------
commit: None
python: 3.7.0.final.0
python-bits: 64
OS: Linux
OS-release: 4.18.3-arch1-1-ARCH
machine: x86_64
processor: 
byteorder: little
LC_ALL: None
LANG: en_US.UTF-8
LOCALE: en_US.UTF-8

pandas: 0.23.4
pytest: None
pip: 10.0.1
setuptools: 40.0.0
Cython: None
numpy: 1.15.0
scipy: 1.1.0
pyarrow: None
xarray: None
IPython: 6.5.0
sphinx: None
patsy: 0.5.0
dateutil: 2.7.3
pytz: 2018.5
blosc: None
bottleneck: None
tables: 3.4.4
numexpr: 2.6.7
feather: None
matplotlib: 2.2.3
openpyxl: None
xlrd: None
xlwt: None
xlsxwriter: None
lxml: None
bs4: None
html5lib: 1.0.1
sqlalchemy: None
pymysql: None
psycopg2: None
jinja2: 2.10
s3fs: None
fastparquet: None
pandas_gbq: None
pandas_datareader: None

</details>

### PR DESCRIPTION
**Pull request by ghost**
closes #14274
closes #19354
closes #22509 
- [ ] tests added / passed
- [X] whatsnew entry

This is a long-standing bug. Effectively, it makes `transform('rank')` synonymous with `g.rank()`  and same for other transformations.

We could also resolve this by restricting the `g.transform(name)` whitelist to only callables and aggs-to-be-broadcasted, telling users to use the named method directly, i.e `g.rank()` instead. Not a problem keeping this path working though, it just needs to return correct results.

Also, continuing some cleanups after #27467

### CODE REVIEW

File: doc/source/whatsnew/v1.0.0.rst [hunk 1]
```diff
@@ -78,6 +78,7 @@ Performance improvements
 Bug fixes
 ~~~~~~~~~
 
+- Previously, when :class:`pandas.core.groupby.GroupBy.transform` was passed the name of a transformation (e.g. `rank`, `ffill`, etc') it  broadcast the results resulting in erroneous values.  only reduction results and not the results of transformations. The return value in thse cases is the same as calling the named function directly (e.g. `g.rank()`) (:issue:`14274`) (:issue:`19354`) (:issue:`22509`).
```
Comment (jbrockmendel) [inline doc/source/whatsnew/v1.0.0.rst]: only -> Only
thse -> these
extra ' after "etc"
Not sure about 1 space vs 2 after a period, but definitely be consistent.

File: pandas/core/groupby/generic.py [hunk 2]
```diff
@@ -615,7 +615,11 @@ def _transform_fast(self, result, obj, func_nm):
         ids, _, ngroup = self.grouper.group_info
         output = []
         for i, _ in enumerate(result.columns):
-            res = algorithms.take_1d(result.iloc[:, i].values, ids)
+            if func_nm in base.reduction_kernels:
+                # only broadcast results if we performed a reduction
+                res = algorithms.take_1d(result.iloc[:, i].values, ids)
```
Comment (jbrockmendel) [inline pandas/core/groupby/generic.py]: would ._values here work instead of .values?  generally less lossy

**Comment by WillAyd** [comment]
Might have gotten lost in a conversation but should these just raise? I think these muddle the API a bit for transform.

I thought you suggested raising for non-reducing items going through agg seems like that could apply here as well

**Comment by pilkibun** [comment]
> I thought you suggested raising for non-reducing items going through agg seems like that could apply here as well

I suggested agg should raise for transformations (will submit it next). That's not the same as *transform* raising for transformations. I thought this route would be less disruptive. 

I'd be perfectly ok with restricting `g.transform` to callables and `agg-to-be-broadcasted` only, that's reasonable, Just as long as there's agreement on this and it doesn't trigger a tug-of-war that stalls this indefinitely. `g.transform('rank')` has been returning nonsense for years and I want to finally fix it *right now*.

File: doc/source/whatsnew/v1.0.0.rst (hunk 1 above)
Comment (pilkibun) [reply doc/source/whatsnew/v1.0.0.rst]: sorry, this was sloppy.

File: pandas/core/groupby/generic.py (hunk 2 above)
Comment (pilkibun) [reply pandas/core/groupby/generic.py]: Interestingly, the SeriesGroupby equivalent already does. Changed.

**Comment by WillAyd** [comment]
Right I'd be +1 for that restriction. In my head I would imagine that we could use the agg whitelist to validate what gets sent to both `.agg` and `.transform`

**Comment by pilkibun** [comment]
Thank you for reviewing

**Comment by pilkibun** [comment]
>  I would imagine that we could use the agg whitelist to validate what gets sent to both .agg and 
> .transform

Did you see the lists added in #27467? I did that so I could do both.

**Comment by WillAyd** [comment]
Yea that's what I was loosely referring to. Let's see what @jreback thinks

File: doc/source/whatsnew/v1.0.0.rst [hunk 3]
```diff
@@ -78,6 +78,7 @@ Performance improvements
 Bug fixes
 ~~~~~~~~~
 
+- Previously, when :class:`pandas.core.groupby.GroupBy.transform` was passed the name of a transformation (e.g. `rank`, `ffill`, etc) it broadcast the results resulting in erroneous values. Only reduction results and not the results of transformations. The return value in these cases is the same as calling the named function directly (e.g. `g.rank()`) (:issue:`14274`) (:issue:`19354`) (:issue:`22509`).
```
Comment (jreback) [inline doc/source/whatsnew/v1.0.0.rst]: this would need a full sub-section with example of previous and new, as this actually is a non-trivial api breaking change. (so put in that section)

**Review by jreback — CHANGES_REQUESTED** [review]
I agree that a non-transformation function (either directly, e.g. rank or an aggregation which broadcasts) should raise.

File: doc/source/whatsnew/v1.0.0.rst (hunk 3 above)
Comment (pilkibun) [reply doc/source/whatsnew/v1.0.0.rst]: Breaking change or bug? I don't understand how broadcasting the result of `rank` could have ever been intended behavior. I can understand `g.transform('rank')` being an alias for `g.rank()`, and I can understand `g.transform('rank')` raising and telling the user to use `g.rank()` directly (as as suggested), but I don't see how any other behavior could be intended. 

Please correct me if I'm wrong.

Comment (pilkibun) [reply doc/source/whatsnew/v1.0.0.rst]: Or are you agreeing its bug, and just saying this is a big deal and we should feature this fix more prominently?

Comment (WillAyd) [reply doc/source/whatsnew/v1.0.0.rst]: It's both but sends a clearer message if put in the API Breaking Changes section

Comment (pilkibun) [reply doc/source/whatsnew/v1.0.0.rst]: Sorry, I misunderstood. Raising on transformations is definitely a breaking change. Will do.

**Comment by pilkibun** [comment]
Ok. Passing the name of anything not in the list of reductions raises an exception. Further, if the method exists on Grouper, the user is instructed to call the method directly.

I also created a section under breaking changes with Previous/New behavior examples.

File: doc/source/whatsnew/v1.0.0.rst [hunk 4]
```diff
@@ -39,8 +39,59 @@ Backwards incompatible API changes
 
 .. _whatsnew_1000.api.other:
 
-- :class:`pandas.core.groupby.GroupBy.transform` now raises on invalid operation names (:issue:`27489`).
--
+Groupby.transform(name) accepts only aggregation names
```
Comment (TomAugspurger) [inline doc/source/whatsnew/v1.0.0.rst]: At a glance, this makes it sound like `.transform(callable)` isn't allowed anymore. Perhaps we can say GroupBy.transform now validates aggregation names.

File: doc/source/whatsnew/v1.0.0.rst [hunk 5]
```diff
@@ -39,8 +39,59 @@ Backwards incompatible API changes
...
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+:meth:`DataFrameGroupBy.transform` and :meth:`SeriesGroupBy.transform` are used to either invoked
```
Comment (TomAugspurger) [inline doc/source/whatsnew/v1.0.0.rst]: This seems a bit wordy. Can we simplify?  For example, I think this entire first sentence can be removed.

The key point to convey is that GroupBy.transform now ensures that the name provided is an aggregation, correct? Anything else?

**Review by TomAugspurger — COMMENTED** [review]
It's not a huge deal, but I think that this could reasonably be called a bug fix. Is there any case where `.transform(name)` was doing the "expected" thing that we'll now raise on?

File: doc/source/whatsnew/v1.0.0.rst (hunk 5 above)
Comment (pilkibun) [reply doc/source/whatsnew/v1.0.0.rst]: My feeling is that there is some confusion as to what `transform` is `supposed` to do. For example, in the issue I opened to discuss agg vs. transform, I understood you to say you you found the broadcasting behavior of `transform('rank')` useful. 

So, I wanted to give a direct explicit description of what the role of `transform` is and that was the purpose of the wordy first sentence, which I've removed now.

File: doc/source/whatsnew/v1.0.0.rst (hunk 4 above)
Comment (pilkibun) [reply doc/source/whatsnew/v1.0.0.rst]: The body explicitly states the callable behavior isn't changed. Changed the title to avoid confusion.

**Comment by pilkibun** [comment]
> was doing the "expected" thing that we'll now raise on?

all the cythonized `cum*` functions took a direct path which was not affected. None of the are aggs, so will now raise.

**Comment by TomAugspurger** [comment]
> all the cythonized cum* functions took a direct path which was not affected. They are all transformations, not aggs, and will now raise.

Can you give an example?

**Comment by pilkibun** [comment]
of what?

**Comment by TomAugspurger** [comment]
Does transform(‘cumsum’) raise?

If so, why? That seems OK to me.

**Comment by pilkibun** [comment]
all the cum* functions are transformations, not aggs.

**Comment by TomAugspurger** [comment]
-1 on outright breaking that then.

And I would need to reread the issue to understand the motivation for changing it.

File: doc/source/whatsnew/v1.0.0.rst [hunk 6]
```diff
@@ -39,8 +39,58 @@ Backwards incompatible API changes
...
 .. _whatsnew_1000.api.other:
 
-- :class:`pandas.core.groupby.GroupBy.transform` now raises on invalid operation names (:issue:`27489`).
--
+Groupby.transform(str) validates name is an aggregation
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+In previous releases, :meth:`DataFrameGroupBy.transform` and
+:meth:`SeriesGroupBy.transform`  did not validate that the function name
+passed was actually the name of an aggregation. As a result, users might get a
```
Comment (jreback) [inline doc/source/whatsnew/v1.0.0.rst]: remove the: As a result.... sentence.

rised -> raise

File: doc/source/whatsnew/v1.0.0.rst [hunk 7]
```diff
@@ -39,8 +39,58 @@ Backwards incompatible API changes
...
+
+    df = pd.DataFrame([0, 1, 100, 99])
+    labels = [0, 0, 1, 1]
+    g = df.groupby(labels)
+
+*Previous behavior*:
+
+.. code-block:: ipython
+
+    In [1]: g.transform('ers >= Decepticons')
```
Comment (jreback) [inline doc/source/whatsnew/v1.0.0.rst]: just pass it a name like 'foo'

**Comment by pilkibun** [comment]
oh my god.

https://github.com/pandas-dev/pandas/pull/27597#issuecomment-515268209

File: doc/source/whatsnew/v1.0.0.rst [hunk 8]
```diff
@@ -39,8 +39,58 @@ Backwards incompatible API changes
...
+    1  2.0
+    2  2.0
+    3  1.0
+
+*New behavior*:
+
+.. ipython:: python
+   :okexcept:
+
+   g.transform('ers >= Decepticons')
```
Comment (jreback) [inline doc/source/whatsnew/v1.0.0.rst]: use foo, & make this a code-block (so we don't have the long traceback)

put the 'rank' in its own ipython block; I would also show .rank() or at least indicate that they are now the same.

File: pandas/core/groupby/groupby.py [hunk 9]
```diff
@@ -241,8 +241,9 @@ class providing the base-class of operations.
 
 Parameters
 ----------
-f : function
-    Function to apply to each group
+func : callable or str
```
Comment (jreback) [inline pandas/core/groupby/groupby.py]: leave this as f, otherwise this is an api change

File: pandas/tests/groupby/test_transform.py [hunk 10]
```diff
@@ -1052,6 +1051,20 @@ def test_transform_agg_by_name(reduction_func, obj):
     assert len(set(DataFrame(result).iloc[-3:, -1])) == 1
 
 
+def test_transform_transformation_by_name(transformation_func):
+    """Make sure g.transform('name') raises a helpful error for non-agg
```
Comment (jreback) [inline pandas/tests/groupby/test_transform.py]: can you add the issue refences numbers as a comment

**Review by jreback — CHANGES_REQUESTED** [review]
can you add tests for the ffill issue that  you are closing

**Comment by TomAugspurger** [comment]
Is https://github.com/pandas-dev/pandas/pull/27597#issuecomment-515598580 answering why “cumsum” should raise?

There’s value in accepting non-agg transform names here. We have an open issue about accepting lists of functions in transform. That change would prohibit .transform([“cummin”, “cummax”]). And using .transform to signal to your reader that the operation is indeed a transform is valuable as well.

**Comment by pilkibun** [comment]
> Is #27597 (comment) answering why “cumsum” should raise?

Tom, I asked you about this exact issue in https://github.com/pandas-dev/pandas/issues/27389 *two weeks ago*. In response you said you found the bug useful and then then proceeded to ignore my followup. Then you showed up here, after I've written the patch, and the tests, and the docs, skipped over the preceding discussion with your peers, ignored the fact that two of them *asked me* to make the change you're objecting to, and casually expect me to rewrite this a third time.
 
I think that's awful behavior, and it's very effectively undermining my desire to help improve pandas.

If this is what I have to put up with in order to *volunteer* my time to fix serious, long-standing bugs,   it's a bad deal and my answer is no.

**Comment by TomAugspurger** [comment]
>  In response you said you found the bug useful 

What comment is that? In https://github.com/pandas-dev/pandas/issues/27389#issuecomment-511396436 I said that I find "broadcasting a transformation result useful in some other cases?". I'm not sure anymore what that's referring to since the post may have been edited.

> and then then proceeded to ignore my followup. 

Don't attribute to malice what can be attributed to busyness.

> Then you showed up here, after I've written the patch, and the tests, and the docs, skipped over the preceding discussion with your peers, ignored the fact that two of them asked me to make the change you're objecting to, and casually expect me to rewrite this a third time.

If I find a change that I think is detrimental to pandas then yes, I'm going to speak up about it. I do apologize if that means lost effort, but I care more about the project as a whole.
